updatehub-sdk/src/listener.rs

```diff
@@ -0,0 +1,140 @@
+    io::BufReader,
+    os::unix::net::{UnixListener, UnixStream},
+    prelude::*,
+};
+use std::{collections::HashMap, net::Shutdown, process};
+
+const SDK_TRIGGER_FILENAME: &str =
+    "/usr/share/updatehub/state-change-callbacks.d/10-updatehub-sdk-statechange-trigger";
+const SOCKET_PATH: &str = "updatehub-statechange.sock";
+
+type StateChangeCallback = fn(State);
+type ErrorCallback = fn(String);
+
+pub struct StateChangeListener {
+    listeners: HashMap<String, StateChangeCallback>,
+    error_handlers: Vec<ErrorCallback>,
+}
+
+pub enum StateID {
+    Download,
+    Install,
+    Reboot,
+    Error,
+}
+
+pub struct State {
+    id: StateID,
+    conn: UnixStream,
+}
+
+impl State {
+    pub fn downloading(&mut self) {
+        self.conn.write_all(b"downloading");
+    }
+
+    pub fn rebooting(&mut self) {
+        self.conn.write_all(b"rebooting");
+    }
+
+    pub fn cancel(&mut self) {
+        self.conn.write_all(b"cancel");
+    }
+
+    pub fn proceed() {}
+}
+
+impl StateChangeListener {
+    pub fn new() -> Self {
+        StateChangeListener { listeners: HashMap::new(), error_handlers: Vec::new() }
+    }
+
+    pub fn on(&mut self, state: StateID, f: fn(State)) {
+        let name = match state {
+            StateID::Download => String::from("download"),
+            StateID::Install => String::from("install"),
+            StateID::Reboot => String::from("reboot"),
+            StateID::Error => String::from("error"),
+        };
+        self.listeners.insert(name, f);
+    }
+
+    fn on_error(&mut self, f: fn(String)) {
+        self.error_handlers.push(f);
+    }
+
+    fn emit(&self, conn: UnixStream, state: &str) {
+        self.listeners.iter().for_each(|(_, f)| {
+            let conn = conn.clone();
+            f(State {
+                id: match state {
+                    "download" => StateID::Download,
+                    "install" => StateID::Install,
+                    "reboot" => StateID::Reboot,
+                    "error" => StateID::Error,
+                    _ => panic!("error"),
+                },
+                conn,
+            })
+        });
+    }
+
+    fn throw_error(&self, error: &str) {
+        self.error_handlers.iter().for_each(|f| {
+            let error = &error;
+            f(error.to_string())
+        })
+    }
+
+    async fn listen(&self) -> Result<()> {
+        let file = std::path::Path::new(&SDK_TRIGGER_FILENAME);
+        let listener = if file.exists() {
+            UnixListener::bind(format!("/run/{}", SOCKET_PATH)).await?
+        } else {
+            println!(
+                "WARNING: updatehub-sdk-statechange-trigger not found on {}",
+                SDK_TRIGGER_FILENAME
+            );
+            UnixListener::bind(SOCKET_PATH).await?
+        };
+
+        let ctrlc = CtrlC::new()?;
+        ctrlc
+            .race(async {
+                loop {
+                    process::exit(1);
+                }
+            })
+            .await;
```

Comment:
This implementation is not correct. The `CtrlC` future will resolve once the signal is received, racing it to a `process::exit(1)` in a loop won't execute as we intend here. Something like this is what you should be doing:
```Rust
let ctrlc_future = async {
  ctrlc.await;
  // handle any necessary socket cleanup/drop
  process::exit(1);
};

let socket_listener_future = async {
  // handle socket listening
};

socket_listener_future.race(ctrlc_future); 
```